### PR TITLE
feat(probes): add serialize-environment warning detection

### PR DESCRIPTION
**Pull request by intincrab**
#359

### CODE REVIEW

**Comment by PierreDemailly** [comment]
Could you add documentation (`/docs/serialize-environment.md`) please ?

File: workspaces/js-x-ray/test/probes/isSerializeEnv.spec.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+test("should not detect other JSON.stringify calls", () => {
+  const str = "JSON.stringify({ foo: 'bar' })";
+  const ast = parseScript(str);
+  const sastAnalysis = getSastAnalysis(str, isSerializeEnv).execute(ast.body);
+
+  assert.strictEqual(sastAnalysis.warnings().length, 0);
+});
+
+test("should not detect non-JSON.stringify calls", () => {
```
Comment (clemgbld) [inline workspaces/js-x-ray/test/probes/isSerializeEnv.spec.js:50]: I wonder if we should add a test that check that edge case like that are handled:
```js
const env = process.env;
JSON.stringify(env);
```

File: workspaces/js-x-ray/src/probes/isSerializeEnv.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+ * JSON.stringify(process["env"])
+ * JSON.stringify(process[`env`])
+ */
+function validateNode(node) {
+  if (node.type !== "CallExpression") {
+    return [false];
+  }
+
+  if (
+    node.callee.type !== "MemberExpression" ||
```
Comment (clemgbld) [inline workspaces/js-x-ray/src/probes/isSerializeEnv.js:18]: I guess you can use the `getMemberExpressionIdentifer` from `@nodesecure/estree-ast-utils`.
it will handle edge case such as:

```js
JSON["s" + "tringify"](process.env);
```

File: workspaces/js-x-ray/src/probes/isSerializeEnv.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    node.callee.object.name !== "JSON" ||
+    node.callee.property.name !== "stringify"
+  ) {
+    return [false];
+  }
```
Comment (fraxken) [inline workspaces/js-x-ray/src/probes/isSerializeEnv.js]: You can probably use utils `getCallExpressionIdentifier` here

File: workspaces/js-x-ray/src/probes/isSerializeEnv.js [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+  if (
+    firstArg.type === "MemberExpression" &&
+    firstArg.object.name === "process" &&
+    (
+      // Check for process.env
+      (firstArg.property.type === "Identifier" && firstArg.property.name === "env") ||
+      // Check for process["env"] or process["env"] or process[`env`]
+      (firstArg.property.type === "Literal" && firstArg.property.value === "env")
+    )
+  ) {
```
Comment (fraxken) [inline workspaces/js-x-ray/src/probes/isSerializeEnv.js]: You can probably also use helper `getMemberExpressionIdentifier`

File: workspaces/js-x-ray/test/probes/isSerializeEnv.spec.js (hunk 1 above)
Comment (fraxken) [reply workspaces/js-x-ray/test/probes/isSerializeEnv.spec.js:50]: I guess we can do it in a second PR by adding initialize and trace what's required

**Comment by fraxken** [comment]
Note: you also need to update README.md

**Review by fraxken — APPROVED** [review]
